modules/core/src/test/java/org/apache/ignite/marshaller/MarshallerEnumDeadlockMultiJvmTest.java

```diff
@@ -0,0 +1,202 @@
+        }
+    }
+
+    /** */
+    private static class BinaryMarshallerFactory implements Factory<Marshaller> {
+        @Override public Marshaller create() {
+            return new BinaryMarshaller();
+        }
+    }
+
+    /** */
+    private static class JdkMarshallerFactory implements Factory<Marshaller> {
+        @Override public Marshaller create() {
+            return new JdkMarshaller();
+        }
+    }
+
+    /**
+     * Attempts to unmarshal both in-built and inner-class enum values at exactly the same time in multiple threads.
+     */
+    private static class UnmarshalCallable implements IgniteCallable<Object> {
+        /** */
+        private final byte[] one;
+        /** */
+        private final byte[] two;
+        /** */
+        @IgniteInstanceResource
+        private Ignite ign;
+
+        /** */
+        public UnmarshalCallable(byte[] one, byte[] two) {
+            this.one = one;
+            this.two = two;
+        }
+
+        /** {@inheritDoc} */
+        @Override public Object call() throws Exception {
+            List<Thread> threadList = new ArrayList<>();
+
+            final CyclicBarrier b1 = new CyclicBarrier(4);
+            final CyclicBarrier b2 = new CyclicBarrier(5);
+
+            for (int i = 0; i < 4; i++) {
+                final int ii = i;
+                Thread thread = new Thread(new Runnable() {
+                    @Override public void run() {
+                        try {
+                            b1.await();
+
+                            if (ii % 2 == 0)
+                                ign.configuration().getMarshaller().unmarshal(one, Thread.currentThread().getContextClassLoader());
+                            else
+                                ign.configuration().getMarshaller().unmarshal(two, Thread.currentThread().getContextClassLoader());
+
+                            b2.await();
+                        }
+                        catch (Exception e) {
+                            e.printStackTrace();
+                        }
+                    }
+                });
+
+                thread.start();
+
+                threadList.add(thread);
+            }
+
+            try {
+                b2.await(5, TimeUnit.SECONDS);
+            }
+            catch (TimeoutException te) {
+                throw new IllegalStateException("Failed to wait for completion", te);
+            }
+
+            for (Thread thread : threadList)
+                thread.join();
+
+            return null;
+        }
+    }
+
+    /** */
+    public enum DeclaredBodyEnum {
+        ONE,
+        TWO {
+            /** {@inheritDoc} */
+            @Override public boolean isSupported() {
+                return false;
+            }
+        };
+
+        /**
+         * Delays initialization slightly to increase chance of catching race condition.
+         */
+        DeclaredBodyEnum() {
+            AtomicInteger cnt = new AtomicInteger();
```

Comment:
I think we definitely shouldn't Thread.sleep in clinit. This can have a whole range of side-effects, such as unlocking the class loader. I think busy wait is the way to go.